crates/core/component/dex/src/component/position_manager/base_liquidity_index.rs

```diff
@@ -0,0 +1,167 @@
+    ///
+    /// # Diagram
+    ///                                                                     
+    ///    Liquidity index:                                                 
+    ///    For an asset `A`, surface asset                                  
+    ///    `B` with the best liquidity                                      
+    ///    score.                                                           
+    ///                             ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐     
+    ///                                                                     
+    ///           ┌──┐              ▼            ┌─────────┐          │     
+    ///     ▲     │  │    ┌──────────────────┐   │         │                
+    ///     │     │ ─┼───▶│{asset_A}{agg_liq}│──▶│{asset_B}│          │     
+    ///     │     ├──┤    └──────────────────┘   │         │                
+    ///   sorted  │  │                           └─────────┘          │     
+    ///   by agg  │  │                                                      
+    ///    liq    ├──┤                                                │     
+    ///     │     │  │                                           used in the
+    ///     │     ├──┤                                            composite
+    ///     │     │  │                                               key    
+    ///     │     │  │       Auxiliary look-up index:                 │     
+    ///     │     │  │       "Find the aggregate liquidity                 
+    ///     │     │  │       per directed trading pair"               │      
+    ///     │     │  │       ┌───────┐                           ┌─────────┐
+    ///     │     │  │       ├───────┤  ┌──────────────────┐     │         │
+    ///     │     │  │       │   ────┼─▶│{asset_A}{asset_B}│────▶│{agg_liq}│
+    ///     │     ├──┤       ├───────┤  └──────────────────┘     │         │
+    ///     │     │  │       ├───────┤                           └─────────┘
+    ///     │     │  │       ├───────┤                                      
+    ///     │     │  │       ├───────┤                                      
+    ///     │     ├──┤       └───────┘                                      
+    ///     │     │  │                                                      
+    ///     │     │  │                                                      
+    ///     │     └──┘                                                      
+    async fn update_asset_by_base_liquidity_index(
+        &mut self,
+        prev_state: &Option<Position>,
+        new_state: &Position,
+        id: &position::Id,
+    ) -> Result<()> {
+        match prev_state {
+            Some(prev_state) => match (prev_state.state, new_state.state) {
+                // We only want to update the index when we process active positions.
+                (Opened, Closed) => {}
+                (Opened, Opened) => {}
+                _ => return Ok(()),
+            },
+            None => {}
+        }
```

Comment:
Not sure this logic is correct. We do want to update the index when a new position is added, no? Can we refactor this into a match statement like
```rust
match (prev_state.map(|prev| prev.state), new_state.state) {
    ...
}
```
or (if it's cleaner) eliminate that match statement entirely in favor of something where we compute two values:

- a previous contribution to the total, to be subtracted (this is 0 if the position is newly opened)
- a current contribution to the total, to be added (this is 0 if the position is newly closed)

if both are 0 it can short circuit without state changes

Reply:
I reworked the structure based on your feedback, now we compute contributions upfront:
- compute `(prev_a, prev_b)` and clamps each value to zero if the position is closed/withdrawn
- compute `(new_a, new_b)` and clamp each value to zero if the position is closed/withdrawn
- short-circuit if all values are zero (e.g. `Closed -> Closed` or other "inactive" transitions)